include/NovelRT/SceneGraph/SceneNode.h

```diff
@@ -0,0 +1,255 @@
+// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root for more information.
+
+#ifndef NOVELRT_SCENENODE_H
+#define NOVELRT_SCENENODE_H
+
+#ifndef NOVELRT_H
+#error Please do not include this directly. Use the centralised header (NovelRT.h) instead!
+#endif
+
+namespace NovelRT::SceneGraph {
+  template <typename T>
+  class breadth_first_traversal_result_iterator {
+  private:
+    std::function<T(const std::shared_ptr<SceneNode>&)> _function;
+    std::queue<std::shared_ptr<SceneNode>> _pendingNodes;
+    std::set<std::shared_ptr<SceneNode>> _visitedNodes;
+
+    T _value;
+
+  public:
+    using iterator_category = std::output_iterator_tag;
+    using value_type = T;
+    using pointer = value_type*;
+    using const_pointer = const value_type*;
+    using reference = value_type&;
+    using const_reference = const value_type&;
+
+    breadth_first_traversal_result_iterator(const std::shared_ptr<SceneNode>& node, std::function<T(const std::shared_ptr<SceneNode>&)> function) :
+      _function(function) {
+      _pendingNodes.push(node);
+      _visitedNodes.insert(node);
+      ++*this;
+    }
+
+    const_reference operator*() const {
+      return _value;
+    }
+
+    const_pointer operator->() const {
+      return std::pointer_traits<pointer>::pointer_to(**this);
+    }
+
+    breadth_first_traversal_result_iterator& operator++() {
+      auto node = _pendingNodes.front();
+      _value = _function(node);
+      _pendingNodes.pop();
+
+      for (auto child : node->getChildren()) {
+        if (_visitedNodes.find(child) == _visitedNodes.cend()) {
+          _visitedNodes.insert(child);
+          _pendingNodes.push(child);
+        }
+      }
+
+      return *this;
+    }
+
+    breadth_first_traversal_result_iterator operator++(int) {
```

Comment:
What's with the naming convention here (and similar places)?

Reply:
is just matching the convention for stdlib iterator names
I couldn't come up with a better one (just maybe `BreadthFirstTraversalResultIterator` or something)
the iterators in particular, users shouldn't be creating themselves